### ISSUE TITLE
Warn on all slice methods that are panicky

### ISSUE THREAD

**Issue by sffc** [rust-lang/rust-clippy#10024]
### What it does

This would be like `clippy::indexing_slicing` except that it would complain about _all_ methods that are panicky based on an index out of bounds, not just the indexing operator.

Examples of methods that would be captured:

- `slice::split_at` and friends like `slice::split_at_mut`, `slice::split_array_ref`, ...
- `slice::rotate_left` and `slice::rotate_right`
- `slice::swap`

### Lint Name

unchecked_index_out_of_bounds

### Category

restriction

### Advantage

- `clippy::indexing_slicing` covers many cases where it's desirable to avoid panics, but it is limited
- From experience reviewing code, index-out-of-bounds is a uniquely common class of errors, more so than other panic conditions
- A generic name like `unchecked_index_out_of_bounds` means that we could easily add more functions to the lint in the future

### Drawbacks

- Slippery slope: where do we draw the line for what panics should be linted?
- `clippy::indexing_slicing` already exists, and perhaps a second, very similar lint would be confusing; should we add these functions to the same lint?

### Example

```rust
let foo = my_slice.split_at(i);
```

Could be written as:

```rust
let foo = my_slice.checked_split_at(i).ok_or(...)?;
```

Note: checked_split_at [doesn't exist yet](https://internals.rust-lang.org/t/slice-split-at-should-have-an-option-variant/17891), so in the mean time, code could be written as

```rust
if i > my_slice.len() {
    return Err(...);
}
#[allow(clippy::unchecked_index_out_of_bounds)] // checked above
let foo = my_slice.split_at(i);
```
